google_compute_engine_oslogin/pam_module/pam_oslogin_login.cc

```diff
@@ -102,10 +109,146 @@ PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
+    pam_syslog(pamh, LOG_ERR,
+               "Bad response from the two-factor start session request: %s",
+               response.empty() ? "empty response" : response.c_str());
+    return PAM_PERM_DENIED;
+  }
+
+  string status;
+  if (!ParseJsonToKey(response, "status", &status)) {
+    pam_syslog(pamh, LOG_ERR,
+               "Failed to parse status from start session response");
+    return PAM_PERM_DENIED;
+  }
+
+  if (status == "NO_AVAILABLE_CHALLENGES") {
+    return PAM_SUCCESS; // User is not two-factor enabled.
+  }
+
+  string session_id;
+  if (!ParseJsonToKey(response, "sessionId", &session_id)) {
+    return PAM_PERM_DENIED;
+  }
+
+  std::vector<oslogin_utils::Challenge> challenges;
+  if (!ParseJsonToChallenges(response, &challenges)) {
+    pam_syslog(pamh, LOG_ERR,
+               "Failed to parse challenge values from JSON response");
+    return PAM_PERM_DENIED;
+  }
+
+  std::map<string,string> user_prompts;
+  user_prompts[AUTHZEN] = "Google phone prompt";
+  user_prompts[TOTP] = "Authenticator app one-time-pass";
+  user_prompts[INTERNAL_TWO_FACTOR] = "Google internal two-factor authentication";
+  user_prompts[IDV_PREREGISTERED_PHONE] = "SMS code";
+
+  oslogin_utils::Challenge challenge;
+  if (challenges.size() > 1) {
+    std::stringstream prompt;
+    prompt << "Available authentication methods: ";
+    for(vector<oslogin_utils::Challenge>::size_type i = 0;
+        i != challenges.size(); ++i)
+      prompt << "\n" << i+1 << ": " << user_prompts[challenges[i].type];
+    prompt << "\n\nEnter a number: ";
+
+    char *choice = NULL;
+    if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &choice, "%s",
+                   prompt.str().c_str()) != PAM_SUCCESS) {
+      pam_error(pamh, "Unable to get user input");
+    }
+
+    int choicei;
+    if (sscanf(choice, "%d", &choicei) == EOF) {
+      pam_error(pamh, "Unable to get user input");
+    }
+    challenge = challenges[choicei - 1];
+  } else {
+    challenge = challenges[0];
+  }
+
+  char* user_token = NULL;
+  if (challenge.type == INTERNAL_TWO_FACTOR) {
+    if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &user_token,
+                   "Enter your security code: ") != PAM_SUCCESS) {
+      pam_error(pamh, "Unable to get user input");
+    }
+  } else if (challenge.type == TOTP) {
+    if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &user_token,
+                   "Enter your one-time password: ") != PAM_SUCCESS) {
+      pam_error(pamh, "Unable to get user input");
+    }
+  } else if (challenge.type == AUTHZEN) {
+    if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &user_token,
+                   "A login prompt has been sent to your enrolled device. "
+                   "Press enter to continue") != PAM_SUCCESS) {
+      pam_error(pamh, "Unable to get user input");
+    }
+  } else if (challenge.type == IDV_PREREGISTERED_PHONE) {
+    if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &user_token,
+                   "A SMS has been sent with a security code. "
```

Comment:
IDV Phone can be sms or voice. Maybe something like "A security code has been sent to your phone."